scipy/optimize/_numdiff.py

```diff
@@ -289,8 +377,62 @@ def fun_wrapped(x):
     return J_transposed.T
 
 
+def _sparse_difference(fun, x0, f0, h, use_one_sided,
+                       structure, groups, method):
+    m = f0.size
+    n = x0.size
+    J = lil_matrix((m, n), dtype=float)
```

Comment:
It might be worth trying to build up row, column, and data arrays and building a COO matrix at the end of the loop, instead of inserting values into a LIL matrix.